kernel: Move block tree db open to block manager

sedited opened the pull request:
Before this change the block tree db was needlessly re-opened during startup when loading a completed snapshot. Improve this by letting the block manager open it on construction. This also simplifies the test code a bit.

The change was initially motivated to make it easier for users of the kernel library to instantiate a BlockManager that may be used to read data from disk without loading the block index into a cache.

sedited committed:
kernel: Move block tree db open to BlockManager constructor

Make the block db open RAII style by calling it in the BlockManager
constructor.

Before this change the block tree db was needlessly re-opened during
startup when loading a completed snapshot. Improve this by letting the
block manager open it on construction. This also simplifies the test
code a bit.

The change was initially motivated to make it easier for users of the
kernel library to instantiate a BlockManager that may be used to read
data from disk without loading the block index into a cache.

josibake commented on src/bitcoin-chainstate.cpp:
```diff
@@ -119,14 +123,12 @@ int main(int argc, char* argv[])
         .blocks_dir = abs_datadir / "blocks",
         .notifications = chainman_opts.notifications,
+        .block_tree_db_dir = abs_datadir / "blocks" / "index",
+        .block_tree_db_cache_size = 2 << 20,
```
I'm guessing you moved the values up from below (L109-L112) so that you could set the value once and use `cache_sizes.block_tree_db` to set `block_tree_db_cache_size` on L127? Otherwise it's not clear to me why those values were moved.

josibake replied:
Sorry, I don't think my last comment on this was very clear. Lines L109-L112 are moved up from below, and `cache_sizes.block_tree_db` is set. So I think L127 should be:

```suggestion
        .block_tree_db_cache_size = cache_sizes.block_tree_db,
```

unless I'm missing something?

josibake commented on src/init.cpp:
```diff
@@ -1536,6 +1531,15 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
     }
     LogPrintf("* Using %.1f MiB for chain state database\n", cache_sizes.coins_db * (1.0 / 1024 / 1024));
 
+    BlockManager::Options blockman_opts{
+        .chainparams = chainman_opts.chainparams,
+        .blocks_dir = args.GetBlocksDirPath(),
+        .notifications = chainman_opts.notifications,
+        .block_tree_db_dir = args.GetDataDirNet() / "blocks" / "index",
+        .block_tree_db_cache_size = cache_sizes.block_tree_db,
+    };
+    Assert(ApplyArgsManOptions(args, blockman_opts)); // no error can happen, already checked in AppInitParameterInteraction
```
Note to self (and other reviewers): this move wasn't immediately obvious to me, until I noticed `CacheSizes` is set on L1521, and then the value `cache_sizes.block_tree_db` is used when creating the `BlockManager::Options`. As a nice side effect, it seems better to have the logs for cache configuration first, and then proceed with setting and creating the `BlockManager`.

josibake commented on src/init.cpp:
```diff
@@ -1559,8 +1563,12 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
         try {
             node.chainman = std::make_unique<ChainstateManager>(*Assert(node.shutdown), chainman_opts, blockman_opts);
+        } catch (dbwrapper_error&) {
+            return InitError(Untranslated("Error opening block database."));
```
Might be premature at this stage, but in theory we could be opening both the block database and txdb at this stage. I haven't looked at `dbwrapper_error` yet so not sure if this is possible, but perhaps we could bubble up the actual db name that failed to open instead of assuming it's always the block database?

TheCharlatan replied:
Heh, we currently always inform the user with "Error opening block database." even if there is an error with the chainstate. I think I am going to leave this as is here, so not to scope creep this PR and will address this in another PR.

josibake commented on src/kernel/blockmanager_opts.h:
```diff
@@ -27,6 +28,14 @@ struct BlockManagerOpts {
     const fs::path blocks_dir;
     Notifications& notifications;
+    const fs::path block_tree_db_dir;
+    // Whether to wipe the block tree database when loading it. If set, this
+    // will also set a reindexing flag so any existing block data files will be
+    // scanned and added to the database.
+    bool wipe_block_tree_db{false};
```
Assuming this is just a comment explaining what was already happening and documenting a behaviour change introduced in this PR? (note to self to test this).

josibake replied:
Nevermind, I can see this a move only from `src/chainstate.h`

josibake commented on src/kernel/chainstatemanager_opts.h:
```diff
@@ -42,7 +42,6 @@ struct ChainstateManagerOpts {
     std::optional<uint256> assumed_valid_block{};
     //! If the tip is older than this, the node is considered to be in initial block download.
     std::chrono::seconds max_tip_age{DEFAULT_MAX_TIP_AGE};
-    DBOptions block_tree_db{};
```
Nice! Makes a lot more sense to me to have block tree database options in the `BlockManagerOptions`

josibake commented on src/node/chainstate.cpp:
```diff
@@ -37,26 +37,6 @@ static ChainstateLoadResult CompleteChainstateInitialization(
     const CacheSizes& cache_sizes,
     const ChainstateLoadOptions& options) EXCLUSIVE_LOCKS_REQUIRED(::cs_main)
 {
-    auto& pblocktree{chainman.m_blockman.m_block_tree_db};
-    // new BlockTreeDB tries to delete the existing file, which
-    // fails if it's still open from the previous loop. Close it first:
-    pblocktree.reset();
```
This code feels a bit smelly, so nice to see you are dropping it by moving the database initialisation into the `BlockManager`. Will need to look at the old code more closely to understand what "previous loop" this comment is referring to.

TheCharlatan replied:
The previous loop this refers to is the initialization retry loop that I am getting rid of in  #30968.

josibake reviewed (commented):
Concept ACK

Conceptually, this makes a lot more sense and makes the code much easier to understand, IMO. Left some comments while doing an initial pass, more questions than concrete suggestions. I did find it a bit hard to parse what was going with a lot of the code being move-only, but from across different files. Not sure if much can be done here, but it would make this easier to follow if you could separate as much of the move-only code changes into their own commit(s)? Specifically, I'm thinking of two instances where the option setting code is moved up a few lines so the cache size option can be passed through.

mzumsande commented on src/init.cpp:
```diff
@@ -1559,8 +1563,12 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
         try {
             node.chainman = std::make_unique<ChainstateManager>(*Assert(node.shutdown), chainman_opts, blockman_opts);
+        } catch (dbwrapper_error&) {
+            return InitError(Untranslated("Error opening block database."));
```
This changes the error handling if the block tree db is corrupted a bit worse as a side effect. Before, we'd get the advice "Please restart with -reindex or -reindex-chainstate to recover." on the console, and, if using the GUI, the option to reindex with by clicking a button. Now we just get a "Error opening block database" error on both without any advice how to fix it.

I'm not sure if the GUI one-click reindex option is totally essential, but in my opinion we should at least keep the advice to reindex, and also show it in the GUI.

josibake replied:
I think this error handling makes more sense post #30968.

I rebased this commit on top of #30968 and was able to get the same error handling behavior as before: https://github.com/josibake/bitcoin/commit/85895b870d669175f42d0730d58930bea169e221, was quite happy with the result!

TheCharlatan replied:
Ugh, I should have drafted this PR once I opened #30968 , since I opened it after being fed up with the init logic in this change. The error handling introduced there forces you to do the right thing in this change now. Sorry for the churn!

mzumsande reviewed (commented):
Concept ACK

TheCharlatan commented:
Rebased 63dfc6067bd7fe06e5a440c229cca030bba6d53a -> 63dfc6067bd7fe06e5a440c229cca030bba6d53a ([blockmanDB](https://github.com/TheCharlatan/bitcoin/tree/blockmanDB) -> [blockmanDB](https://github.com/TheCharlatan/bitcoin/tree/blockmanDB), [compare](https://github.com/TheCharlatan/bitcoin/compare/blockmanDB..blockmanDB))

* Fixed conflict with #30968

TheCharlatan commented:
Updated 63dfc6067bd7fe06e5a440c229cca030bba6d53a -> 60ade81516d42d275c1143f49f47e142d32c45fc ([blockmanDB_1](https://github.com/TheCharlatan/bitcoin/tree/blockmanDB_1) -> [blockmanDB_2](https://github.com/TheCharlatan/bitcoin/tree/blockmanDB_2), [compare](https://github.com/TheCharlatan/bitcoin/compare/blockmanDB_1..blockmanDB_2))

* Addressed @josibake's [comment](https://github.com/bitcoin/bitcoin/pull/30965#discussion_r1789888213), reusing a option value in `bitcoin-chainstate`
* Exchanged use of `Untranslated` with `_`, since the string in question is displayed to GUI users.

josibake commented:
crACK https://github.com/bitcoin/bitcoin/commit/60ade81516d42d275c1143f49f47e142d32c45fc

Verified that the change is mostly move-only. Makes the code much more clear and cleanly separates initialising / opening the database from loading the database contents into the in memory cache.

> The change was initially motivated to make it easier for users of the kernel library to instantiate a BlockManager that may be used to read data from disk without loading the block index into a cache.

Will test this and report back if I find anything unexpected!

mzumsande reviewed (commented):
Conceptual question:

I think that just the goal of not loading the blocktreedb twice could have been achieved easier by simply moving the code from `CompleteChainstateInitialization` into `LoadChainstate()`.

You also list other motivations, but I don't completely get these:
It seems a little bit non-intuitive to me to have things like Cleanup of blk / rev files (`CleanupBlockRevFiles`) done in the constructor, instead of having a separate call for this.
It also doesn't feel consistent if the blocktreedb is loaded automatically when creating a `ChainstateManager`, but the coinsb of the chainstates aren't - I would have expected both or neither. 

So can you explain a bit more in detail how this helps kernel?

TheCharlatan commented:
Re https://github.com/bitcoin/bitcoin/pull/30965#pullrequestreview-2352892509

Thank you for the questions @mzumsande!

> So can you explain a bit more in detail how this helps kernel?

I will try to give a thought-dump of what I am attempting to achieve here. As said in the title description, the kernel-motivated goal of this PR is to allow users of the library to instantiate the block tree db in the block manager in the same way we would currently in the chainstate load routines without having to load everything into the cache or having to load a chainstate. This allows a potential user to immediately use the block tree db, for example to recurse through it and read ordered blocks from disk.

As an alternative to the approach here, opening the database could happen in a separate routine, outside of the `BlockManager` constructor, but from my current mental model, this would then require our current code to either check if this routine was called before, or force the user to close it again before doing other tasks. I would consider this a less elegant solution.

>  It seems a little bit non-intuitive to me to have things like Cleanup of blk / rev files (CleanupBlockRevFiles) done in the constructor, instead of having a separate call for this.

My thinking here is that since this is always the correct thing to do if the corresponding options are set, having a separate call is not really useful. If certain options are set and I instantiate a `BlockManager` I would expect it to be in the state I configured it with, without having the call additional functions.

> It also doesn't feel consistent if the blocktreedb is loaded automatically when creating a ChainstateManager, but the coinsb of the chainstates aren't - I would have expected both or neither.

Yes, my eventual goal was to move calling `InitializeChainstate` for the fully validated chainstate into the `ChainstateManager` constructor as well as calling `InitCoinsDB` for it. I wanted to do this in a future pull request if this change is accepted.

TheCharlatan commented:
Rebased 60ade81516d42d275c1143f49f47e142d32c45fc -> 1c1df66cebbe825a24cb2f66a24aab78709db7db ([blockmanDB_2](https://github.com/TheCharlatan/bitcoin/tree/blockmanDB_2) -> [blockmanDB_3](https://github.com/TheCharlatan/bitcoin/tree/blockmanDB_3), [compare](https://github.com/TheCharlatan/bitcoin/compare/blockmanDB_2..blockmanDB_3))

* Fixed conflict with #30967

TheCharlatan commented:
Rebased f55d36a19c61e0ec8c06a6e5a0bbae455c8a1d06 -> f55d36a19c61e0ec8c06a6e5a0bbae455c8a1d06 ([blockmanDB](https://github.com/TheCharlatan/bitcoin/tree/blockmanDB) -> [blockmanDB](https://github.com/TheCharlatan/bitcoin/tree/blockmanDB), [compare](https://github.com/TheCharlatan/bitcoin/compare/blockmanDB..blockmanDB))

mzumsande reviewed (commented):
Thanks - your explanation makes sense - I did a bit of research on how much logic and I/O in constructors are considered to be "good style", and there doesn't really seem to be a clear consensus to me.

For me the main downside would be that some of the things the ctor does may be unexpected, hence below suggestion to add some documentation.

TheCharlatan commented:
Thank you for the review @mzumsande,

Updated f55d36a19c61e0ec8c06a6e5a0bbae455c8a1d06 -> 182945e946ac6a92daad012d6579e5441b9641cc ([blockmanDB_4](https://github.com/TheCharlatan/bitcoin/tree/blockmanDB_4) -> [blockmanDB_5](https://github.com/TheCharlatan/bitcoin/tree/blockmanDB_5), [compare](https://github.com/TheCharlatan/bitcoin/compare/blockmanDB_4..blockmanDB_5))

* Addressed @mzumsande's [comment](https://github.com/bitcoin/bitcoin/pull/30965#discussion_r1901165977), added documentation to chainman creation specifying what the constructor does.
* Addressed @mzumsande's [comment](https://github.com/bitcoin/bitcoin/pull/30965#discussion_r1901166334), removed now unused `cache_sizes` param in `CompleteChainstateInitialization`.

TheCharlatan commented:
Re https://github.com/bitcoin/bitcoin/pull/30965#pullrequestreview-2521245377

>  I did a bit of research on how much logic and I/O in constructors are considered to be "good style", and there doesn't really seem to be a clear consensus to me.

I wish there would be better consensus on what the best practices are. If somebody has reasoning for doing things differently, do share!

ryanofsky commented on src/bitcoin-chainstate.cpp:
```diff
@@ -119,14 +123,12 @@ int main(int argc, char* argv[])
         .chainparams = chainman_opts.chainparams,
         .blocks_dir = abs_datadir / "blocks",
         .notifications = chainman_opts.notifications,
+        .block_tree_db_dir = abs_datadir / "blocks" / "index",
+        .block_tree_db_cache_size = cache_sizes.block_tree_db,
```
In commit "kernel: Move block tree db open to block manager" (182945e946ac6a92daad012d6579e5441b9641cc)

It seems far from ideal for bitcoin-chainstate and other kernel library users (not to mention our own test code) to need to hardcode the "blocks/index" path and to manually copy the cache size from one struct to another. I think a better API would allow these option fields to be unset and internally use a `Flatten` method similar to the ones in `validation.cpp` and `txmempool.cpp` to set appropriate default values when the fields are unset.

TheCharlatan replied:
I like the suggestion, but after implementing it still does not seem entirely ideal. I just made the entire `DBParams` field an optional, so it's not easily possible to set only the cache size, but not the path. It is also annoying that the current behaviour with the path is that it is dependent on the data dir path, not the blocks dir path.

ryanofsky reviewed (approved):
Code review ACK 182945e946ac6a92daad012d6579e5441b9641cc. I think a number of improvements are possible here (see comments below) but it does make sense to move block database initialization out of `src/node/chainstate.cpp` and into the `BlockManager` class, and move its options out of `ChainstateManager::Options` and into `BlockManager::Options`.

On constructor initialization question I do think it's often useful if lightweight construction with delayed initialization or lazy initialization is possible, but it should be very easy to support these things if needed by just splitting the constructor up and overloading it.

sedited committed:
refactor: Remove redundant reindex check

The check for whether the block tree db has been wiped before calling
NeedsRedownload() is confusing. The boolean is set in case of a reindex.
It was originally introduced to guard NeedsRedownload in case of a
reindex in #21009. However NeedsRedownload already returns early if the
chain's tip is not loaded. Since that is the case during a reindex, the
pre-check is redundant.

mzumsande commented on src/node/chainstate.cpp:
```diff
@@ -152,7 +124,7 @@ static ChainstateLoadResult CompleteChainstateInitialization(
     }
 
-    if (!options.wipe_block_tree_db) {
+    if (!chainman.m_blockman.m_opts.wipe_block_tree_db) {
```
I agree that it should be possible to drop the check.

This is a off-topic here, but maybe the entire `NeedsRedownload()`  check could be removed at some point, or  simplified to just check block one after segwit, instead over iterating over 400k indexes each startup? Segwit activation is now more than 8 years in the past after all...

TheCharlatan commented:
Updated 182945e946ac6a92daad012d6579e5441b9641cc -> 384c73d4fcda4af7c2b4bfb945a248cb93dba47d ([blockmanDB_5](https://github.com/TheCharlatan/bitcoin/tree/blockmanDB_5) -> [blockmanDB_6](https://github.com/TheCharlatan/bitcoin/tree/blockmanDB_6), [compare](https://github.com/TheCharlatan/bitcoin/compare/blockmanDB_5..blockmanDB_6))

* Addressed @ryanofsky's [comment](https://github.com/bitcoin/bitcoin/pull/30965#discussion_r1906052767), added back interrupt check.
* Addressed @ryanofsky's [comment](https://github.com/bitcoin/bitcoin/pull/30965#discussion_r1906064655), added a commit for dropping the if statement in question.
* Addressed @ryanofsky's [comment](https://github.com/bitcoin/bitcoin/pull/30965#discussion_r1906073567), consolidated individual fields into a single DBParams field.

ryanofsky commented on src/init.cpp:
```diff
@@ -1198,18 +1202,33 @@ static ChainstateLoadResult InitAndLoadChainstate(
+    // The coinsdb is opened at a later point on LoadChainstate.
     try {
         node.chainman = std::make_unique<ChainstateManager>(*Assert(node.shutdown_signal), chainman_opts, blockman_opts);
+    } catch (dbwrapper_error& e) {
+        LogError("%s\n", e.what());
+        return {ChainstateLoadStatus::FAILURE, _("Error opening block database")};
```
In commit "refactor: Remove redundant reindex check" (fd590d5ca1cfb6c2525b7755edb6a3a2cf16c856)

Curious about something: In https://github.com/bitcoin/bitcoin/pull/30965#issuecomment-2397922175 you wrote,

> Yes, my eventual goal was to move calling `InitializeChainstate` for the fully validated chainstate into the `ChainstateManager` constructor as well as calling `InitCoinsDB` for it. I wanted to do this in a future pull request if this change is accepted.

So if coins db were intialized in the chainstate constructor, then this dbwrapper_error exception could mean there is an error either from the coins db or the blocks db, not just the blocks db as stated in the current error message.

For this reason, and in any case, it seems fragile for this code to have to know what ChainstateManager constructor is doing internally and what exceptions that code might throw. I think it might be better if instead of letting dbwrapper_error be thrown from the ChainstateManager constructor, code in that constructor or code in the BlockStorage constructor would catch that exception and throw a more generic exception that this code could handle without needing to know about ChainstateManager internals.

Maybe a good candidate for a generic exception type could be the existing `ChainstateLoadResult` type since it can hold an error message and also indicate whether the error is fatal or not (i.e. whether a reindex should be attempted or not).

TheCharlatan replied:
Good thinking ahead! Yeah, my idea was to use something along those lines too. Maybe we could start annotating some of the functions here to make it easier for the developer to see which exceptions it throws? Though my understanding is that such annotations have been deliberately left out of the code so far (https://stackoverflow.com/questions/1055387/throw-keyword-in-functions-signature).

On the other hand it would be nice if we could agree on more coherent error handling for constructors, be it with factory functions, error out params, or exceptions. I am starting to like exceptions for these cases though, and combining them with `ChainstateLoadResult` would make responding to them fairly straight forward in our init code.

ryanofsky commented on src/node/chainstate.cpp:
```diff
@@ -34,35 +34,9 @@ namespace node {
-    }
-
-    if (chainman.m_interrupt) return {ChainstateLoadStatus::INTERRUPTED, {}};
+   if (chainman.m_interrupt) return {ChainstateLoadStatus::INTERRUPTED, {}};
```
In commit "kernel: Move block tree db open to block manager" (384c73d4fcda4af7c2b4bfb945a248cb93dba47d)

Looks like indentation unintentionally changed here

ryanofsky reviewed (approved):
Code review ACK 384c73d4fcda4af7c2b4bfb945a248cb93dba47d but it seems like a fuzz check is failing in CI. Since last review, BlockManagerOpts uses DBParams, so no longer needs to duplicate most of its fields, and blockstorage m_opts no longer need to be public so some other changes could be reverted. One other set of changes that were reverted had do to with `m_block_tree_db_in_memory` / `opts.block_tree_db_in_memory` in tests, and seem more correct now.

TheCharlatan commented:
Updated 384c73d4fcda4af7c2b4bfb945a248cb93dba47d -> b776e40554c8a315d832f3996d14ff2e3ae7b8cb ([blockmanDB_6](https://github.com/TheCharlatan/bitcoin/tree/blockmanDB_6) -> [blockmanDB_7](https://github.com/TheCharlatan/bitcoin/tree/blockmanDB_7), [compare](https://github.com/TheCharlatan/bitcoin/compare/blockmanDB_6..blockmanDB_7))

* Addressed @ryanofsky's [comment](https://github.com/bitcoin/bitcoin/pull/30965#discussion_r1909124834), ran clang-format over commits.

mzumsande reviewed (commented):
Code Review ACK b776e40554c8a315d832f3996d14ff2e3ae7b8cb

ryanofsky reviewed (approved):
Code review ACK b776e40554c8a315d832f3996d14ff2e3ae7b8cb. Just whitespace changes since last review

TheCharlatan commented:
Rebased b776e40554c8a315d832f3996d14ff2e3ae7b8cb -> 7f484b319bdc9e646e270b6d6d9a1c7e52296dc0 ([blockmanDB_7](https://github.com/TheCharlatan/bitcoin/tree/blockmanDB_7) -> [blockmanDB_8](https://github.com/TheCharlatan/bitcoin/tree/blockmanDB_8), [compare](https://github.com/TheCharlatan/bitcoin/compare/blockmanDB_7..blockmanDB_8))

* Resolved conflict with #31483

TheCharlatan commented on src/init.cpp:
```diff
@@ -1199,18 +1203,33 @@ static ChainstateLoadResult InitAndLoadChainstate(
         .notifications = chainman_opts.notifications,
+        .block_tree_db_params = DBParams{
+            .path = args.GetDataDirNet() / "blocks" / "index",
+            .cache_bytes = static_cast<size_t>(cache_sizes.block_tree_db),
```
yes -_-

TheCharlatan commented:
Updated 7f484b319bdc9e646e270b6d6d9a1c7e52296dc0 -> f3cfdf2b511776211a0399510ced735e556d510d ([blockmanDB_8](https://github.com/TheCharlatan/bitcoin/tree/blockmanDB_8) -> [blockmanDB_9](https://github.com/TheCharlatan/bitcoin/tree/blockmanDB_9), [compare](https://github.com/TheCharlatan/bitcoin/compare/blockmanDB_8..blockmanDB_9))

* Drop `static_casts` as pointed out by @maflcko's [comment](https://github.com/bitcoin/bitcoin/pull/30965#discussion_r1922387437).

maflcko reviewed (approved):
left some style-nits (only if you re-touch)

review ACK f3cfdf2b511776211a0399510ced735e556d510d 🥃

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: review ACK f3cfdf2b511776211a0399510ced735e556d510d 🥃
B0REdECuAQLOEWL1/F7/sHyOzZiaGn5sptzUNU7w1dEMjVCfDkqW5G60UcFwChqEIxDmcB0qfNeBXz1nC0DlDw==
```

</details>

sedited committed:
kernel: Move block tree db open to block manager

This commit is done in preparation for the next commit. Here, the block
tree options are moved to the blockmanager options and the block tree is
instantiated through a helper method of the BlockManager, which is
removed again in the next commit.

Co-authored-by: MarcoFalke <*~=`'#}+{/-|&$^_@721217.xyz>

TheCharlatan commented:
Updated f3cfdf2b511776211a0399510ced735e556d510d -> 0cdddeb2240d1f33c8b2dd28bb0c9d84d9420e3d ([blockmanDB_9](https://github.com/TheCharlatan/bitcoin/tree/blockmanDB_9) -> [blockmanDB_10](https://github.com/TheCharlatan/bitcoin/tree/blockmanDB_10), [compare](https://github.com/TheCharlatan/bitcoin/compare/blockmanDB_9..blockmanDB_10))

* Took @maflcko's [suggestion](https://github.com/bitcoin/bitcoin/pull/30965#issuecomment-2602793391), for splitting move-only changes into a separate commit.
* Addressed @maflcko's [comment](https://github.com/bitcoin/bitcoin/pull/30965#discussion_r1922589316), remove no-op lock.
* Addressed @maflcko's [comment](https://github.com/bitcoin/bitcoin/pull/30965#discussion_r1922596559), remove newline on touched `LogError`.

mzumsande reviewed (commented):
re-ACK 0cdddeb2240d1f33c8b2dd28bb0c9d84d9420e3d

achow101 commented:
ACK 0cdddeb2240d1f33c8b2dd28bb0c9d84d9420e3d

maflcko reviewed (commented):
sorry for missing the bug in review
